src/backend/distributed/utils/background_jobs.c

```diff
@@ -0,0 +1,1251 @@
+}
+
+
+#endif
+
+
+/*
+ * bgw_generate_returned_message -
+ *      generates the message to be inserted into the job_run_details table
+ *      first part is comming from error_severity (elog.c)
+ */
+static void
+bgw_generate_returned_message(StringInfoData *display_msg, ErrorData edata)
+{
+	const char *prefix = error_severity(edata.elevel);
+	appendStringInfo(display_msg, "%s: %s", prefix, edata.message);
+	if (edata.detail != NULL)
+	{
+		appendStringInfo(display_msg, "\nDETAIL: %s", edata.detail);
+	}
+
+	if (edata.hint != NULL)
+	{
+		appendStringInfo(display_msg, "\nHINT: %s", edata.hint);
+	}
+
+	if (edata.context != NULL)
+	{
+		appendStringInfo(display_msg, "\nCONTEXT: %s", edata.context);
+	}
+}
+
+
+/*
+ * UpdateDependingTasks updates all depending tasks, based on the type of termnial state
+ * the current task reached.
+ */
+static void
+UpdateDependingTasks(BackgroundTask *task)
+{
+	switch (task->status)
+	{
+		case BACKGROUND_TASK_STATUS_DONE:
+		{
+			UnblockDependingBackgroundTasks(task->jobid, task->taskid);
+			break;
+		}
+
+		case BACKGROUND_TASK_STATUS_ERROR:
+		{
+			/* when we error this task, we need to unschedule all dependant tasks */
+			UnscheduleDependantTasks(task->jobid, task->taskid);
+			break;
+		}
+
+		default:
+		{
+			/* nothing to do for other states */
+			break;
+		}
+	}
+}
+
+
+/*
+ * ConsumeTaskWorkerOutput consumes the output of an executor and mutates the
+ * BackgroundTask object to reflect changes like the message and status on the task.
+ */
+static void
+ConsumeTaskWorkerOutput(shm_mq_handle *responseq, BackgroundTask *task, bool *hadError,
+						bool *cancelled)
+{
+	/*
+	 * Message-parsing routines operate on a null-terminated StringInfo,
+	 * so we must construct one.
+	 */
+	StringInfoData msg = { 0 };
+	initStringInfo(&msg);
+
+	for (;;)
+	{
+		resetStringInfo(&msg);
+
+		/*
+		 * Get next message. Currently blocking, hen multiple backends get implemented it
+		 * should switch to a non-blocking receive
+		 */
+		Size nbytes = 0;
+		void *data = NULL;
+		const bool noWait = false;
+		shm_mq_result res = shm_mq_receive(responseq, &nbytes, &data, noWait);
+
+		if (res != SHM_MQ_SUCCESS)
+		{
+			break;
+		}
+
+		appendBinaryStringInfo(&msg, data, nbytes);
+
+		char msgtype = pq_getmsgbyte(&msg);
+		switch (msgtype)
+		{
+			case 'E': /* ERROR */
+			{
+				if (hadError)
+				{
+					*hadError = true;
+				}
+				__attribute__((fallthrough));
+			}
+
+			case 'N': /* NOTICE */
+			{
+				ErrorData edata = { 0 };
+				StringInfoData display_msg = { 0 };
+
+				pq_parse_errornotice(&msg, &edata);
+				initStringInfo(&display_msg);
+				bgw_generate_returned_message(&display_msg, edata);
+
+				if (edata.sqlerrcode == ERRCODE_QUERY_CANCELED)
+				{
+					if (cancelled)
+					{
+						*cancelled = true;
+					}
+				}
+
+				StringInfoData fullMessage = { 0 };
+				initStringInfo(&fullMessage);
+				if (task->message)
+				{
+					appendStringInfoString(&fullMessage, task->message);
```

Comment:
I noticed we only save the last one in pg_cron, however, I was thinking it would be nicer to see all. Although it does make sense that some things could be very chatty.

Wonder if that is higher risk with pg_cron, compared to this, since it will be very limited amount of systems that would use this. That said, that might be an assumption that would be proven wrong in the near future.

Reply:
I have reverted back to only storing the last message.

Contrary to pg_cron, I do append the tag returned by the process. So this will be included with the last message.